index.bs

```diff
@@ -68,12 +71,145 @@ The appropriate times to <a spec=HTML>fetch and process the linked resource</a>
+    1. [=list/Append=] |rule| to |result|'s [=speculation rule set/prefetch rules=].
+  1. Return |result|.
+</div>
+
+<div algorithm="parse a speculation rule">
+  To <dfn export>parse a speculation rule</dfn> given a [=map=] |input| and a [=URL=] |baseURL|, perform the following steps. They return a [=speculation rule=] or null.
+
+  1. If |input| has any [=map/key=] other than "`source`", "`urls`", and "`requires`", then return null.
+  1. If |input|["`source`"] does not [=map/exist=] or is not the [=string=] "`list`", then return null.
+  1. Let |urls| be an empty [=list=].
+  1. If |input|["`urls`"] does not [=map/exist=], is not a [=list=], or has any element which is not a [=string=], then return null.
+  1. [=list/For each=] |urlString| of |input|["`urls`"]:
+    1. Let |parsedURL| be the result of [=basic URL parser|parsing=] |urlString| with |baseURL|.
+    1. If |parsedURL| is failure, then [=iteration/continue=].
+    1. If |parsedURL|'s [=url/scheme=] is not an [=HTTP(S) scheme=], then [=iteration/continue=].
+    1. [=list/Append=] |parsedURL| to |urls|.
+  1. Let |requirements| be an empty [=ordered set=].
+  1. If |input|["`requires`"] [=map/exists=], but is not a [=list=], then return null.
+  1. [=list/For each=] |requirement| of |input|["`requires`"]:
+    1. If |requirement| is not the [=string=] "`anonymous-client-ip`", then return null.
+    1. [=set/Append=] |requirement| to |requirements|.
+  1. Return a [=speculation rule=] with [=speculation rule/URLs=] |urls| and [=speculation rule/requirements=] |requirements|.
+</div>
+
+<h3 id="speculation-rules-processing">Processing Model</h3>
+
+A [=document=] has a <dfn for=document export>list of speculation rule sets</dfn>, which is an initially empty [=list=].
+
+<!-- TODO(domfarolino): Get rid of the `data-link-type="interface"` once we fix the dfn in HTML. -->
+Periodically, the user agent may [=queue a task=] to the <a data-link-type="interface">DOM manipulation task source</a> to [=consider speculation=] for any [=document=].
+
+<p class="note">
+  The user agent will likely do this after the insertion of new speculation rules, or when resources are idle and available.
+
+<div algorithm="consider speculation">
+  To <dfn export>consider speculation</dfn> for a [=document=] |document|:
+
+  1. For each |ruleSet| of |document|'s [=document/list of speculation rule sets=]:
+    1. [=list/For each=] |rule| of |ruleSet|'s [=speculation rule set/prefetch rules=]:
+      1. Let |requiresAnonymousClientIP| be true if |rule|'s [=speculation rule/requirements=] [=set/contains=] "`anonymous-client-ip`", and false otherwise.
+      1. [=list/For each=] |url| of |rule|'s [=speculation rule/URLs=]:
+        1. The user agent MAY prefetch |url| given |requiresAnonymousClientIP|.
+           <p class="issue">TODO: expand this to actually elaborate on how prefetch works, once initiated, and to incorporate the |requiresAnonymousClientIP| flag.
```

Comment:
We should probably remove the speculation rule set from this list after we process, right? Else it seems like we might re-process the same speculation rule set N times for as many speculation rules scripts as a dev provides.

Reply:
Relatedly, should the spec deduplicate multiple prefetches for the same URL? For the same (url, requires anonymous client IP) pair? We could always let UAs do that deduplication via the "MAY", but it seems nicer to nail it down...

Reply:
We certainly could say something like. "The user agent should not prefetch |url| if it has already done so for this document", but it kinda leads into asking how the UA should determine that, which seems like it doesn't make as much sense until the data structures that define prefetching and prefetch matching exist. At that point, maybe we can use "should" language with reference to whatever maps are used for that.

Reply:
Yeah, OK, deferring this until we have an idea of how prefetch works makes sense. Maybe just update the TODO to note this so that we don't forget.

Reply:
Okay, done.